channeldb/payment_control.go

```diff
@@ -148,10 +152,15 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 		// retrying the payment or return a specific error.
 		case err == nil:
 			switch paymentStatus {
-
 			// We allow retrying failed payments.
 			case StatusFailed:
 
+			// We already have payment creation info for this
+			// payment so we won't allow creating a duplicate one.
+			case StatusInitiated:
```

Comment:
Still in my mind the question keeps popping up why a new `StatusInitiated` is needed instead of using `StatusInFlight` for this.

Reply:
Because it's not inflight? In my view they are different states, and in later commits we can see that we allow deleting `StatusInitiated` but not `StatusInFlight`.

Reply:
I am still missing the final goal of adding this state, the difference that it makes to the user. I'd think that it is only in Initiated for such a short time that deleting it isn't that useful?

Reply:
It's a logical state. And yes, it's only for a short time when things go well. Suppose a payment is created, then lnd crashes, you'll be safe to remove it when you see it's in `StatusInitiated`. I think that'll be a useful case?

Reply:
A related question is how to get rid of payments in `StatusInFlight` when lnd crashes right at the time when there are momentarily no pending htlcs.

Reply:
The `Status` field serves as a small but nice abstraction of an MPP's current state, which means, ideally, a user/caller can act based on it solely. On the other hand, we can replace it with `StatusInFlight` plus a check over the htlcs, but that IMO defeats the purpose of having the status in the first place. By the same argument, neither `StatusSucceed` nor `StatusFailed` is needed as we can always check the states of the HTLCs.

There's a `DeletePayment` to delete payments, only if it's not in `StatusInFlight`.

Reply:
I am still not sold on why it is important to invest in this extra state. More code, more questions, and also because you expose it on the rpc level, it ties external applications to what I think is an internal detail.

I have no objection to a payment status that is strictly speaking redundant (`StatusSucceed`). It avoids redetermining that state in multiple places. But functionally speaking, it looks like `StatusInitiated` doesn't add anything over just using `StatusInFlight`, except for the deletion edge case. Why is this a priority to change?

Reply:
> More code, more questions

That's true. Down the road, we will face the question anyway. When updating the payment, adding more htlc attempts, or deleting it, we have to ask, is the payment's current state allows such an action? We can, as you've said, just check by looking into its HTLCs. Or, as you've also mentioned, it avoids determining that state by using the new state.

> I have no objection to a payment status that is strictly speaking redundant (StatusSucceed).

It shouldn't be redundant though, unless you think the payment status field is redundant.


> except for the deletion edge case.

Then how can we solve this issue?

Reply:
>Then how can we solve this issue?

In the same way in which deleting left-over payments after a restart that did have previous htlc attempts are going to be solved?

Reply:
For my understanding: The reason for this new state is so that we can check `removable()` (added later on) and delete payments that have no HLTCs?